macOS Intel: crash in release modes on the C side (LLVM codegen issue?)

### Zig Version

0.9.0-dev.1722+ed70f9981

### Steps to Reproduce

cc @kubkon: here's the promised ticket from https://github.com/ziglang/zig/issues/9747#issuecomment-974805026)

Ok, this is a weird problem, brace yourself:

On an Intel Mac (my config: Big Sur 11.6.2, Xcode 13.1, macOS SDK 12.0, Zig installed via nightly download, but ```brew install --HEAD``` has the same behaviour):

```sh
git clone https://github.com/floooh/sokol-zig
cd sokol-zig
zig build -Drelease-safe run-clear
```
The executable crashes on the C side with:

```
Assertion failed: ((desc->buffer_pool_size > 0) && (desc->buffer_pool_size < _SG_MAX_POOL_SIZE)), function _sg_setup_pools, file sokol_gfx.h, line 13332.
```

Or alternatively:
```
zig build -Drelease-fast run-clear
```
In this case the crash happens when trying to allocate a large piece of memory on the C side (which would have been caught by the assert in release-safe mode):

```
clear(5067,0x116afae00) malloc: can't allocate region
:*** mach_vm_map(size=18446744065119621120, flags: 100) failed (error code=3)
clear(5067,0x116afae00) malloc: *** set a breakpoint in malloc_error_break to debug
```

When buiding without release-safe or release-fast, it works fine.

The problem seems Intel Mac specific, it doesn't happen on M1 Macs or on Windows (haven't tested on Linux).

The problem occurs inside this call:

https://github.com/floooh/sokol-zig/blob/603a4abf693002d0cc2fd1b391ff111f78b698e0/src/examples/clear.zig#L13-L15

The provided structure is default-initialized to zero:

https://github.com/floooh/sokol-zig/blob/603a4abf693002d0cc2fd1b391ff111f78b698e0/src/sokol/gfx.zig#L641-L654

The function call into C goes through this stub where the address of the struct that's passed by value in the Zig call is taken when calling into the C side:

https://github.com/floooh/sokol-zig/blob/603a4abf693002d0cc2fd1b391ff111f78b698e0/src/sokol/gfx.zig#L655-L658

Eventually the call ends up here on the C side:

https://github.com/floooh/sokol-zig/blob/603a4abf693002d0cc2fd1b391ff111f78b698e0/src/sokol/c/sokol_gfx.h#L14773

Adding debug-printfs on the Zig- and C-side reveals that the structure content *doesn't* get corrupted on its way from Zig to C (which was my first suspicion), instead the breakage happens in this C code block which "patches" default values into zero-initialized struct members:

https://github.com/floooh/sokol-zig/blob/603a4abf693002d0cc2fd1b391ff111f78b698e0/src/sokol/c/sokol_gfx.h#L14790-L14800

Those ```_sg_def``` macros resolve to:
```
_sg.desc.buffer_pool_size = (((_sg.desc.buffer_pool_size) == 0) ? (default_value) : (_sg.desc.buffer_pool_size));
```

Interestingly, adding debug-printf-calls which print _sg.desc.buffer_pool_size before that construct makes the problem go away (great example of a Heisenbug I guess).

I can workaround the problem by **not** using the same left-hand-side and right-hand-side-values:
```
_sg.desc.buffer_pool_size = (((desc->buffer_pool_size) == 0) ? (default_value) : (desc->buffer_pool_size));
```

No idea how to proceed from here. It looks to me like perfectly valid C code. Is this a side effect of the last LLVM update perhaps or different compilation flags for C code? But I don't see the same problem with Clang in C projects, I'm also using -fstrict-aliasing for my C/C++ code anywhere.

### Expected Behavior

...

### Actual Behavior

...

Ok, I hit a little snag with bisecting. The 'status' changes between these two versions from beginning of October:

```
2021-10-01 05:59     https://ziglang.org/builds/zig-macos-x86_64-0.9.0-dev.1217+c4cd592f0.tar.xz
2021-10-02 01:41     https://ziglang.org/builds/zig-macos-x86_64-0.9.0-dev.1244+7350cb2fc.tar.xz
```

Version 1244 already has the crash. The problem is I can't say if version 1217 is a "good" version because I'm currently hitting a macOS SDK related compilation error (which was fixed between 1217 and 1244 it seems):

```
In file included from /Users/floh/scratch/sokol-zig/src/sokol/c/sokol_gfx.h:2551:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Metal.framework/Headers/Metal.h:9:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Metal.framework/Headers/MTLTypes.h:8:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:12:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSBundle.h:91:143: error: function does not return NSString
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:103:48: note: expanded from macro 'NS_FORMAT_ARGUMENT'

Compile C Objects [6/7] sokol_app.c... error(compilation): clang failed with stderr: In file included from /Users/floh/scratch/sokol-zig/src/sokol/c/sokol_app.c:5:
```

I'll see if I can get further back if I revert form Xcode 13.x to Xcode 12.x.

--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -4996,17 +4996,9 @@ fn writeSymbolTable(self: *MachO) !void {
         }
     }
 
-    var undefs = std.ArrayList(macho.nlist_64).init(self.base.allocator);
-    defer undefs.deinit();
-
-    for (self.undefs.items) |sym| {
-        if (sym.n_strx == 0) continue;
-        try undefs.append(sym);
-    }
-
     const nlocals = locals.items.len;
     const nexports = self.globals.items.len;
-    const nundefs = undefs.items.len;
+    const nundefs = self.undefs.items.len;
 
     const locals_off = symtab.symoff;
     const locals_size = nlocals * @sizeOf(macho.nlist_64);
@@ -5021,7 +5013,7 @@ fn writeSymbolTable(self: *MachO) !void {
     const undefs_off = exports_off + exports_size;
     const undefs_size = nundefs * @sizeOf(macho.nlist_64);
     log.debug("writing undefined symbols from 0x{x} to 0x{x}", .{ undefs_off, undefs_size + undefs_off });
-    try self.base.file.?.pwriteAll(mem.sliceAsBytes(undefs.items), undefs_off);
+    try self.base.file.?.pwriteAll(mem.sliceAsBytes(self.undefs.items), undefs_off);
 
     symtab.nsyms = @intCast(u32, nlocals + nexports + nundefs);
     seg.inner.filesize += locals_size + exports_size + undefs_size;

--- a/src/link/MachO/Atom.zig
+++ b/src/link/MachO/Atom.zig
@@ -437,9 +437,11 @@ pub fn parseRelocs(self: *Atom, relocs: []macho.relocation_info, context: RelocC
                         };
                         addend = mem.readIntLittle(i32, self.code.items[offset..][0..4]) + correction;
                         if (rel.r_extern == 0) {
+                            // Note for the future self: when r_extern == 0, we should subtract correction from the
+                            // addend.
                             const seg = context.object.load_commands.items[context.object.segment_cmd_index.?].Segment;
                             const target_sect_base_addr = seg.sections.items[rel.r_symbolnum - 1].addr;
-                            addend += @intCast(i64, context.base_addr + offset + correction + 4) -
+                            addend += @intCast(i64, context.base_addr + offset + 4) -
                                 @intCast(i64, target_sect_base_addr);
                         }
                     },

